Enabling AQE on [databricks]

NVnavkumar opened the pull request:
Fixes #1059 

This branch fixes a couple of issues with enabling Spark Adaptive Query Execution (AQE) on the Databricks Spark environment.  Currently this is marked as WIP since I'm still investigating whether there are any more obvious ongoing issues with enabling adaptive execution in Databricks systems. Here are the issues fixed so far:

1) implementing certain method calls that are currently required by the Databricks environment

2) implementing handling of a logical plan window optimization that happens in Databricks distributions but not currently in Apache Spark.

3) Fixed the Databricks shim for GpuShuffleExchangeExec for 10.4 to fix the issue with a duplicate submission of the map stage which causes a missing job id reference.

4) Added to the AQEUtils shim to handle 9.1 so that ShuffleExchangeExec will fallback to CPU when AQE is enabled.

<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

NVnavkumar commented:
build

NVnavkumar commented:
This is actually an in issue in GpuShuffleExchangeExec sending an incorrect job id (maybe stage id?) to the scheduler.  Current workaround is to fallback ShuffleExchangeExec to CPU when using Databricks and AQE.

> Looks like I found a new failure when using SQL UNION, seems to crash the system:
>

NVnavkumar commented:
build

NVnavkumar commented:
Fixed the issue in `GpuShuffleExchangeExec` on Databricks 10.4. For Databricks 9.1, ShuffleExchangeExec will fallback to CPU when AQE is enabled in that environment, due to the complexities of how it handles the submission of the map stage. Updated the lead comment to reflect updates.

revans2 commented:
> Fixed the issue in `GpuShuffleExchangeExec` on Databricks 10.4. For Databricks 9.1, ShuffleExchangeExec will fallback to CPU when AQE is enabled in that environment, due to the complexities of how it handles the submission of the map stage. Updated the lead comment to reflect updates.

Have we measured the performance impact of this?  In many cases AQE is not that big of a performance win, but sending the data to the CPU and back again is a really big performance hit.

NVnavkumar commented:
Note that it is still on the GPU in Databricks 10.4. We can measure the performance impact for the older version which has this fallback if that makes sense.


On Sep 2, 2022, at 7:55 AM, Robert (Bobby) Evans ***@***.***> wrote:

﻿

Fixed the issue in GpuShuffleExchangeExec on Databricks 10.4. For Databricks 9.1, ShuffleExchangeExec will fallback to CPU when AQE is enabled in that environment, due to the complexities of how it handles the submission of the map stage. Updated the lead comment to reflect updates.

Have we measured the performance impact of this? In many cases AQE is not that big of a performance win, but sending the data to the CPU and back again is a really big performance hit.

—
Reply to this email directly, view it on GitHub<https://github.com/NVIDIA/spark-rapids/pull/6461#issuecomment-1235600115>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AXFDIM3CFV2WRLAYCW33BADV4IIMXANCNFSM6AAAAAAQA4JWIA>.
You are receiving this because you were assigned.Message ID: ***@***.***>

tgravescs commented on sql-plugin/src/main/311+-db/scala/com/nvidia/spark/rapids/shims/ShimLeafExecNode.scala:
```diff
@@ -0,0 +1,37 @@
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExecBase
+
+trait ShimLeafExecNode extends LeafExecNode {
+  override def computeStats(): Statistics = {
```
we should add a few comments, especially just saying Long.MaxValue was recommended

tgravescs commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/AQEUtils.scala:
```diff
@@ -16,18 +16,21 @@
   /** Return a new QueryStageExec reuse instance with updated output attributes */
   def newReuseInstance(sqse: ShuffleQueryStageExec, newOutput: Seq[Attribute]): QueryStageExec = {
     val reusedExchange = ReusedExchangeExec(newOutput, sqse.shuffle)
+    logWarning(s"reusedExchange: $reusedExchange")
```
remnove or make Debug

tgravescs reviewed (commented):
overall seems fine, I thought at one point you had mentioned implementing computeStats per exec and trying to get a realistic size, did that not work out?

NVnavkumar commented:
> overall seems fine, I thought at one point you had mentioned implementing computeStats per exec and trying to get a realistic size, did that not work out?

If we wanted to do it correctly and most accurately, it is a bit an undertaking to implement the PlanVisitor for not just the leaf exec's but also some of the operations. In particular, we have to implement potentially our own Join estimation. Also, when I started looking into it, I got some strangely different numbers in our Parquet reading exec's vs what Databricks is reporting. I figure that can be done separately as potentially a task in the future (if there is any upside to the computation).

NVnavkumar commented:
> > Fixed the issue in `GpuShuffleExchangeExec` on Databricks 10.4. For Databricks 9.1, ShuffleExchangeExec will fallback to CPU when AQE is enabled in that environment, due to the complexities of how it handles the submission of the map stage. Updated the lead comment to reflect updates.
> 
> Have we measured the performance impact of this? In many cases AQE is not that big of a performance win, but sending the data to the CPU and back again is a really big performance hit.

Actually figured out how to get GPU Shuffle back in 9.1, will push an update shortly

NVnavkumar commented:
build

tgravescs commented on integration_tests/src/main/python/aqe_test.py:
```diff
@@ -0,0 +1,145 @@
+# Dynamic coalescing of shuffle partitions
+# Dynamically Handle Skew Joins
+
+_adaptive_coalese_conf = copy_and_update(_adaptive_conf, {
```
are we calling this out explicitly just to make sure enabled everywhere or was this off in one of distributions?  if just to make sure perhaps we want to include the other ones like  spark.sql.adaptive.skewJoin.enabled

NVnavkumar replied:
This was some old code, I think we should handle this configuration testing more so in https://github.com/NVIDIA/spark-rapids/issues/6253. I have removed it for now

tgravescs commented on integration_tests/src/main/python/aqe_test.py:
```diff
@@ -0,0 +1,145 @@
+@ignore_order(local=True)
+def test_aqe_skew_join():
+    def do_join(spark):
+        left, right = create_skew_df(spark, 500)
```
is there anyway for us to really know AQE kicked in here?  The different versions of Spark have changed things so I can see some of these being different or potentially changing and not testing what we meant.  Perhaps we need to enhance our plan checks to make sure AQE is there and it did what we expect.  I'm ok if that is follow on if we need

NVnavkumar replied:
so for many of these tests, it's not as much a question of whether AQE kicked. In Databricks, these are each basically calling out the leaf exec's where computeStats needs to be implemented. That is only called when AQE is enabled on Databricks. I could see them backporting this to future Apache versions though. There is already a scala test version of this one that actually examines the plan. Perhaps I could look into how that one is kicked in?

tgravescs replied:
ok lets just put comments on the ones that are testing computeStats.  

we can do the testing separately with https://github.com/NVIDIA/spark-rapids/issues/6253

tgravescs commented on integration_tests/src/main/python/aqe_test.py:
```diff
@@ -0,0 +1,145 @@
+
+@ignore_order(local=True)
+@pytest.mark.parametrize("data_gen", integral_gens, ids=idfn)
+def test_aqe_join_parquet(spark_tmp_path, data_gen):
```
can we comment on explicitly what each one of these is trying to test

tgravescs commented on integration_tests/src/main/python/aqe_test.py:
```diff
@@ -0,0 +1,145 @@
+@ignore_order(local=True)
+@pytest.mark.parametrize("data_gen", integral_gens, ids=idfn)
+def test_aqe_join_parquet_batch(spark_tmp_path, data_gen):
+    # force v2 source for parquet to use BatchScanExec
```
why?

tgravescs commented on integration_tests/src/main/python/aqe_test.py:
```diff
@@ -0,0 +1,145 @@
+
+@ignore_order
+@pytest.mark.parametrize('data_gen', integral_gens, ids=idfn)
+@pytest.mark.parametrize('aqe_enabled', ['true', 'false'], ids=idfn)
```
why test with aqe off here?

NVnavkumar replied:
I tested this one on and off because of the window changes. I can remove this here (it was useful for development).

tgravescs commented on sql-plugin/src/main/312db/scala/com/nvidia/spark/rapids/shims/AQEUtils.scala:
```diff
@@ -28,6 +28,5 @@ object AQEUtils {
 
-  // currently we don't support AQE on Databricks
-  def isAdaptiveExecutionSupportedInSparkVersion: Boolean = false
+  def isAdaptiveExecutionSupportedInSparkVersion: Boolean = true
```
we explicitly put this in to work around an issue with delta lake on Databricks in 10.4, sync with me so we can make sure we test that issue.  ie this is used in isDeltaLakeMetadataQuery, fi all of these are true now we may be able to remove that code.

tgravescs replied:
Can you file a follow on issue to test and remove this workaround?

NVnavkumar replied:
Filed https://github.com/NVIDIA/spark-rapids/issues/6539

tgravescs commented on sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/AQEUtils.scala:
```diff
@@ -16,18 +16,18 @@
 
 /** Utility methods for manipulating Catalyst classes involved in Adaptive Query Execution */
-object AQEUtils {
+object AQEUtils extends Logging {
```
remove Logging bit

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:
```diff
@@ -340,8 +340,11 @@ object GpuWindowExec extends Arm {
+            // stage, replacing them with aliases. Also, sometimes children are not provided in the
+            // initial list of expressions after optimizations, so we add them here, and they will
+            // be deduped anyways in the other passes
+            val newChildren = wf.children.map(ce =>
```
@mythrocks  mind taking a look at this

mythrocks replied:
Sorry for the delay.

I'm not particularly familiar with the logic behind `isPreNeeded`, etc. But on discussion with @NVnavkumar, one wonders if we should check why `isPreNeeded` is turning up `false` on Databricks, with AQE turned on. We might adjust how `isPreNeeded` is calculated.

NVnavkumar replied:
So, `isPreNeeded` is `false` on Databricks, but it's a bit of a red herring in this case. We actually need to the window function children in the GpuWindowExec itself, it looks like the extra GpuProject does not help in this case.

NVnavkumar replied:
After some exploration, I have determined that the bug this fixes is not AQE-specific and I'm not 100% confident that this fix is currently the right approach. I have reverted this fix and filed a new issue https://github.com/NVIDIA/spark-rapids/issues/6531 to track the bug here.

NVnavkumar commented:
build

NVnavkumar commented:
build

NVnavkumar commented:
build

NVnavkumar commented:
build

NVnavkumar commented:
build
